### PR TITLE
Center panel sync

### PR DESCRIPTION
**Pull request by will-moore**
This PR uses the jsTree for images in the centre panel (everything apart from SPW data).
This means that loading Thumbnails should be MUCH faster!

NB: Not yet ready for full testing but generally working OK.
 - Some things not yet done:
     - Remove old unused code.

Need to test all functionality that affects the centre panel.

- Browsing P/D/I S/P/W in tree - Multi-select of P/D/I, browse between P/D/I and SPW etc.
- Double - click thumbnail to open image
- Cut, copy paste in tree
- Selection of images in centre panel (click, shift/cmd click & drag-select)
- Filtering of images in centre panel (along with selection - E.g. select all then filter - filtered images de-selected)
- Zooming of thumbnails - should be preserved between Datasets.
- Switch between icon / table layouts - should be preserved between Datasets.
- Pagination
- Browsing of Tags / Tagsets (containing P/DI, S/PW).
- Public browsing of shares - Opening shared images with share ID.
- Refresh tree will update thumbnails if available (E.g. new rdef in Insight or Web client in another browser).
- Web integration tests should be green.

### CODE REVIEW

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#103](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/103/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/103/consoleText) for more details.
Possible conflicts:
  - PR #4203 will-moore 'Api link multiple'
    - components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/data/includes/center_plugin.thumbs.js.html

**Comment by dominikl** [comment]
Tested the mentioned workflows with firefox, works well (and fast!). Only slightly irritating thing I noticed: Selecting an image in the tree always refreshes the whole central panel.

**Comment by will-moore** [comment]
I realise that Robot tests (delete image in share) is failing, since I now no-longer show the "Object Deleted" thumbnail https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/113/robot/Web%20&%20Web/Web/Delete%20Test/Test%20Delete%20Images%20in%20Share/

**Comment by will-moore** [comment]
@dominikl Is there a particular workflow that makes the refreshing centre panel on image-selection noticeable? Mostly it shouldn't be obvious that the whole centre is being refreshed, since the html should be the same (apart from the selection itself).

**Comment by dominikl** [comment]
Just selecting an image in the left hand side tree; thumbnails then turned white and got reloaded again. Can't reproduce at the moment, but I guess the PR is not in the build today, is it?

**Comment by will-moore** [comment]
@dominikl That last commit should fix the refreshing of thumbnails on selection changes.

**Comment by pwalczysko** [comment]
Bug:  When using the key to go down on the tree, then if I go over the Datasets, some of them expand as I drive over them, some of them do not. Yet again, the datasets which expand and do not expand might  be the very same, just did expand the first time, do not expand the second time and vice versa. There seem to be no logic in this behaviour. Note that when I mouse-click onto a dataset in the tree, it ALWYAS expands.

**Comment by dominikl** [comment]
Could confirm, unnecessary thumbnail reloading has been fixed.

**Comment by pwalczysko** [comment]
Further info to https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148398241 - once I have driven once over a dataset, and it did expand, if I drive back (say first I was using down key, now I am using up key), it will not expand anymore on the second visit.

**Comment by pwalczysko** [comment]
Bug: Drag and select followed by Shift Click in central pane.

  - have a Dataset with 5 images
  - all in cetral pane, select the middle 3 images by drag and select
  - Shift click onto the first image -> the last image is selected as well -> not expected
  - it does not happen all the time, but definitely repeatable

**Comment by pwalczysko** [comment]
Bug: Central pane does not load in some situations.

  - go to screen http://eel.openmicroscopy.org/merge/webclient/?show=screen-355
  - expand it, expand the plate, click on the Run
  - the wells in central pane load as expected
  - now in the tree, select the screen again
  - then select the Run again
  - observe that this time, the wells in cetral pane do not load anymore, the central plane is empty

**Comment by pwalczysko** [comment]
Bug: 
> Refresh tree should update thumbnails when available....

This test completely failed
  - go to plate ID 1355 on eel
  - in Insight, set all the images to Greyscale and ``Save to All``
  - in Web, navigate to the plate and see that the Rnd settings have been respected -> works as expected but
  - now, with the plate still highlighted in Web, go back to Insight and ``Set Imported and Save``
  - the plate is now red (as expected with imported settings in Insight) but
  - click ``Refresh`` in the tree of Web and 
  - observe that you completely lost the context, you have to navigate to the plate/run/wells manually
  - when you do so, the wells are loaded with the old, black settings (the change made in Insight was not respected) -> bug
  - refresh the whole browser and observe that now the changes from Insight have been applied

**Comment by pwalczysko** [comment]
Similarly to https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148432698 the Run does not load when in Tags tab  and navigating away in the tree to the tag, then back to the Run. Even the refresh of the tree will not reload the central pane with the wells.

**Comment by pwalczysko** [comment]
Bug: Tags tree freezes when navigated to.

  - in Insight, create a Tagset and add several tags into it
  - in web or insight, add some tags from the tagset to several projects
  - in web, select the project in the tree, which has one or more of the tags from the tagset mentioned above
  - in the right-hand pane click on the tag to get to it in the Tags tab
  - the Tags tab loads as expected but
  - the Tagset which contains the tag is not expanded
  - the Tagset which contains the tag cannot be expanded even manually
  - any other tags which also contain images/plates/datasets cannot be expanded either
  - no action in the Tags tab possible
  - this is a regression against eel latest
  - Example Tag ID is on eel merge user-4 http://eel.openmicroscopy.org/merge/webclient/?show=tag-12751

**Comment by pwalczysko** [comment]
Bug: Share thumbnails not visible for share members.

  - as user-3, create a share with user-1 (Share 16858)
  - logout, login as user-1 and navigate to the Share 16858
  - observe that all the thumbnails in the share have a grey placeholder
  - observe that the Full viewer works as expected from the central pane (thumbs), tree and right-hand pane button

screen shot 2015-09-29 at 11 24 36

**Comment by pwalczysko** [comment]
> Public browsing of shares - Opening shared images with share ID.

@will-moore :  What exactly do you mean by this ? Is by this meant a normal Share workflow as tested in https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148479444 ?

**Comment by pwalczysko** [comment]
Bug: Pagination - second page not in sync between central and tree. Problems when switching from second page to first with selections and sync.

   - when object on the second page are selected in central pane, they are not selected in tree at all (this behaviour is okay on the 1st page though)
   - when navigating from the second page back to first page, the click sometimes does not cause the action, a second click onto the ``1`` is necessary
   - when the first page is loaded back, any first time selection in central or tree is not respected in central or tree (out of sync just after reload of first page)
   - expample dataset ``pagination`` has ID  	2602 eel merge user-4

**Comment by pwalczysko** [comment]
Blockers:
  - Share problem https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148479444
  - Freeze in Tags tab when Tagset  https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148477309
  - No respect for Rnd changes in another app https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148468910
  - Central pane does not load in some cases https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148432698, https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148472305 and https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148398241, https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148404268
  - Pagination is broken https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148487604


Problems in selection in central pane:
  - Shift Click for first time and Drag and select problems https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148414634

**Comment by will-moore** [comment]
@pwalczysko Thanks for the thorough testing. 
 - Share bug https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148479444 is fixed
 - Following a link ?show=tag-123 when that tag is under a Tag-set https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148477309 is certainly a bug but not part of this PR.
 - Sorry - should have mentioned that Rendering changes (thumbnail versioning) is not working in SPW data (it never has, even in 5.1). We have a ticket for that somewhere. I tried testing this again for Datasets and it worked for me. Could you give me exact sequence to see the bug in Datasets?
 - Center panel loading for "Run", after selecting a Screen, Tag or other object should be fixed. However, the failure of Datasets to expand when selected a second time by Hot-keys is a bug - will try to fix that later.
 - Haven't had time to look at Pagination or Shift-click bugs yet.

To summarise: fixed share thumbnails and loading of Runs - can be re-tested.  Thanks!

**Comment by will-moore** [comment]
Update - Also fixed the drag-select then shift-click of thumbnails.
In the case where you drag-select a column of thumbnails, then shift-click another thumbnail, the behaviour may not be entirely what you expect, but this is probably OK as you are really mixing modes of selection here and user probably doesn't have a clear idea what to expect anyway. At least when you drag select a contiguous number of thumbnails (a row of thumbnails), subsequent shift-click should work as expected.

**Comment by pwalczysko** [comment]
Drag & select, Shift click, Ctrl click behaviours, 

workflows and combinations retested. 
For me, this fully works as expected now. Even when thumbs from different rows were selected followed by Shift click selection on anything else, the selection goes now towards the Shift clicked image all the way (but just there, not on the other side of the Drag-selected group) and the gaps between the Drag selected images are filled.

Do not see how these workflows could be improved yet, this seems fully satisfactory for me./

**Comment by pwalczysko** [comment]
@will-moore : Rnd settings done in another client on Datasets question - I just cannot repeat the problem on Dataset I had yesterday, maybe there was some caching. Today, on Datasets, everything works as expected. Also confirmed that the Plate Rnd settings change problem is present also in eel latest. This part also seems to be clean now.

**Comment by pwalczysko** [comment]
Re: 
>Following a link ?show=tag-123 when that tag is under a Tag-set #4242 (comment) is certainly a bug but not part of this PR.

But it does not happen on eel latest. When I follow such link (to a tag contained in a tagset), I am confronted with non-expanded Tagset in Tags tab, but the Tagset expansion can easily be done manually afterwards and it works. 
On the other hand, on eel merge, such link causes complete freeze of the ability to expand any tag or tagset in the tags tab. This seems much worse behaviour, and is caused by some open PR.

**Comment by pwalczysko** [comment]
The Share thumbnail problem is fixed in large. But there is one detail to it : The thumbs in Share are a bit distorted. See the comparison of the same data in the Data pane and in the Share (Public pane)

Data pane
screen shot 2015-10-16 at 10 41 24

Share pane
screen shot 2015-10-16 at 10 41 10

**Comment by pwalczysko** [comment]
The Run selection bug is fixed.

**Comment by pwalczysko** [comment]
In summary after today's testing:
  - Share bug is fixed up to some thumb weird shapes https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148668990
  - Run selection bug is fixed (but not the hot key double selection on object as indicated)
  - Shift-click, Drag and select and Ctrl Click behaviours fixed and fully satisfactory
  - Tag link navigation when in Tagset -> not sure that there is no worsening of the behaviour due to this PR, needs one more look
  - Rnd settings changes respecting -> confirmed no regression on Runs, and working fine on Dataset


Seems to me that the ToDos would be:

   - clarify the reason for the distorted Share thumbs (this PR at fault or not), and if not, file a ticket
   - fix the hot key selection/expansion of Dataset problem https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148404268
   - fix the pagination problems https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-148487604

**Comment by will-moore** [comment]
All ToDos should be fixed:
 - Share thumbnails sizes (and also dates and size X, Y, Z in table layout of share) should be fixed.
 - Hotkey selection of Datasets should always load thumbnails, even when re-selected multiple times.
 - Pagination should not break selection in tree when thumbnails selected.
 - Following link to tag that is within a tag-set still won't cause the tree to 'freeze'. (Actually traversing the tree and selecting the tag itself never has worked - probably not time to fix this for 5.2.0).

**Comment by pwalczysko** [comment]
When Projects or Datasets are tagged, then under Tags tab, the central pane shows them (unexpectedly) as inaccesible placeholder thumbnails.

screen shot 2015-10-19 at 10 19 23

**Comment by pwalczysko** [comment]
All the problems are fixed as advertised by @will-moore https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-149048925. The only thing ToDo is the newly found tags tab problem https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-149156169

**Comment by pwalczysko** [comment]
Refresh after Rnd settings change not happening - the thumbs and Preview get out of sync. 
   - Find ``leica-original`` dataset under 5.0 project on eel merge (any user will do)
   - open the first image in Preview and click ``Grayscale``
   - then click ``Save to All``
   - observe that every other image in the MIF becomes grayscale -> expected
   - in the tree, select the containing Dataset (``leica-original``)
   - right-click and select ``Set Imported and Save``
   - observe that now there is a discrepancy between the central pane and the Preview

**Comment by will-moore** [comment]
@pwalczysko That should fix the discrepancy of thumbnails after ```Save to All```. Please also check that Save of a single image is working (just the single thumbnail updates) and that this doesn't then get out of sync with Preview.

**Comment by pwalczysko** [comment]
- ``Save to All`` behaviour okay (sync with Preview, central pane and tree, also with changing selection passed)
  - ``Save`` to single image okay (again, sync tests passed)
  - Listing of Projects as images under Tags tab -> problem fixed


All good here. Ready to merge.

**Comment by pwalczysko** [comment]
Sorry, problems with http://eel.openmicroscopy.org/merge/webclient/?show=dataset-1605 - login as user-1 - scroll to the bottom of the first page and observe error.

**Comment by pwalczysko** [comment]
Cannot repeat https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-149516039.

**Comment by pwalczysko** [comment]
Bug: After multiselect - annotation actions, the selection sync between central and right-hand pane is lost.

  - go to a dataset with couple of images
  - select more than 2 of those images
  - in the right hand pane, add a tag to all of the images
  - now in central pane, select just one of the images
  - observe that the right hand pane still shows the multi-selection from previous step
  - observe that when you click onto the image in tree, the syncing with the right-hand pane is restored

**Comment by will-moore** [comment]
@pwalczysko That last bug should be fixed by that commit(s).

**Comment by pwalczysko** [comment]
The last bug (multiselect action followed by desync in central pane) is fixed.

The https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-149516039 cannot be repeated even today. 
Ready to merge.

**Comment by jburel** [comment]
user-5 read-write
Browsing Kymograph results datasets. No thumbnails for the last image (9790). 
It should be similar to the first one (9784)

**Comment by will-moore** [comment]
@jburel Thumbnail was so thin that height was getting set to 0. Fixed now.

**Comment by jburel** [comment]
@will-moore: thanks

**Comment by jburel** [comment]
maybe to be done in another PR.
Whn the central panel is in "table view"mode, If I select a node in the tree, the viewport in the central panel is not always modified so the node can be visible

**Comment by pwalczysko** [comment]
Bug: (will CERTAINLY cause displeasure of users):

The display of long names in the tree in case of images is not handled well in 5.2. The names when long get "under" the central pane. The UI displays the long names in such a way that just the ends are visible, but in 5.2 "a bit" of the end is hidden on the right, which defeats the purpose of the feature.
No such problem on 5.1 (trout).

eel (5.2):

screen shot 2015-10-22 at 10 24 10


trout (5.1)
screen shot 2015-10-22 at 10 28 47

**Comment by pwalczysko** [comment]
- the Kymograph issue https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-149917700 is fixed
  - cannot find any other problems here except https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-150160802 and https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-150147878 which might be done in another PR

**Comment by will-moore** [comment]
@pwalczysko @jburel OK, I agree it would be best to merge this now, and fix those issues in a new PR.

**Comment by jburel** [comment]
Both comments (mine+Petr) added to https://trello.com/c/6ugBp0HG/5-web-center-panel-follow-up
for 5.2.1

**Comment by will-moore** [comment]
I think @pwalczysko's issue https://github.com/openmicroscopy/openmicroscopy/pull/4242#issuecomment-150160802 needs to be fixed in 5.2.0 since it is a blocker. Will try to open a new PR with it ASAP.

**Comment by jburel** [comment]
Thanks merging
